CIP-0021/CIP-0021.md

```diff
@@ -0,0 +1,104 @@
+
+Another motivation for this CIP is that without canonical CBOR serialisation, specifically canonical map key order, HW wallets would not be able to verify that multi-asset policies and asset_names are included only once in a given output or policy respectively, as they don't have enough memory to store all the previously passed tokens for the sake of deduplication. This might then cause the user to sign a transaction with the same token included twice, although at the protocol level, only one use of the token would be accounted for.
+
+## Specification
+
+This specification is a modification of the canonical CBOR suggestions taken from the [CBOR specification RFC](https://datatracker.ietf.org/doc/html/rfc7049#section-3.9). Parts not mentioned in this CIP should follow the canonical CBOR suggestions from the CBOR RFC.
+
+### Map key ordering
+
+All keys in a map should be ordered based on their real, logical value, not their CBOR representation. Number type keys (uints, ints and floats) should be ordered based on their numerical value. Byte string and text string keys should be ordered based on their lexicographic order. If the keys are of different types then the order described below should be maintained.
+
+**Ordering of keys with different types**
+
+The order of types is based on the CBOR major type’s 3 bit header identificator:
+
+1. unsigned integer - major type 0
+2. negative integer - major type 1
+3. byte string - major type 2
+4. text string - major type 3
+5. floating-point numbers - major type 7
+
+I.e. first come the uints, then the ints, byte strings, text strings and floating-point numbers.
+
+Arrays, maps and tagged items are left out of this specification as including them might unnecessarily complicate the specification or its implementation. Also, these elements aren’t currently used as keys in maps and are unlikely to be used in the future.
+
+**Examples of correctly ordered keys**
+
+- `"aa", "b", "c", "cccc", "d"`
+- `h'0000', h'01', h'02020202', h'03'`
+- `1, 2, 3, 4, 100, 1000`
+- `1, 100, -100, -200, h'01', h'0202020202', "aa", "b", 0.7, 2.4`
+
+## Rationale
+
+### Why not follow the canonical CBOR suggestions completely?
+
+The canonical CBOR suggestions order the keys based on their CBOR representations with the rather weird caveat that shorter keys come first regardless of their lexicographic ordering. Lexicographic ordering is more natural and it might be simpler for many applications to sort based on the real, logical value of the key rather than the serialised representation.
+
+### Keys with mixed types
+
+We mustn’t forget that the CBOR map keys can be of different types. The type ordering can be chosen arbitrarily but it seems most natural to choose the ordering based on the CBOR types’ 3 bit header identificator.
```

Comment:
I don't really feel strongly about any of those arguments. Since there already exists a specification for producing canonical CBOR serialization, coming up with a new one does not sound to me like a good idea? 

The main argument seems to be that it is "most natural to choose ordering based on the CBOR types". It certainly is appealing to the human brain but encoded data are processed by programs anyway. How does the proposal makes it simpler for many applications :thinking: ?